### PR TITLE
Node and Content libs: wrong query results #9850

### PR DESCRIPTION
**Pull request by edloidas**
Types for aggregations and suggestions in result and in params were incorrect. I tried to make them match the Java implementation and work with JS examples, but I'm not 100% everything is correct, so it's better to test them on some real data before actually releasing.

What were changed:
* `StatsAggregation` interface was declared twice and used both in result and in params types. The stats aggregation in result is now called `StatsAggregationResult`.
* `suggestions` in result type was invalid, according to the [examples](https://github.com/enonic/xp/blob/master/modules/lib/lib-node/src/main/resources/lib/xp/examples/node/query.js#L175).
* Same for the `aggregations`. See [examples](https://github.com/enonic/xp/blob/master/modules/lib/lib-node/src/main/resources/lib/xp/examples/node/query.js#L129).

### CODE REVIEW

File: modules/lib/lib-node/src/main/resources/lib/xp/node.ts
Comment (edloidas) [reply modules/lib/lib-node/src/main/resources/lib/xp/node.ts:640]: The reason why I removed `Suggestion` type and placed `TermSuggestion` here was that it was basically an alias for that. Having two same types with different names exported doesn't look very wise. Since we haven't released any stable types yet, it'll be better to export as less as possible if we are not sure how this types will look in future releases. It's always easier to add types, than remove them, because in the second case it can break stuff.

**Comment by ComLock** [comment]
durationStats doesn't appear in the expected, so it's untested?
https://github.com/enonic/xp/blob/master/modules/lib/lib-node/src/main/resources/lib/xp/examples/node/query.js#L66
https://github.com/enonic/xp/blob/master/modules/lib/lib-node/src/main/resources/lib/xp/examples/node/query.js#L110

**Comment by ComLock** [comment]
ASFAIK Stats Aggregations can only be used as a subAggregation, so that might be an issue.

**Review by ComLock — APPROVED** [review]

File: modules/lib/lib-content/src/main/resources/lib/xp/content.ts
Comment (tajakobsen) [inline modules/lib/lib-content/src/main/resources/lib/xp/content.ts:86]: I mentioned earlier that it is a bit difficult to split `AggregationsResult`, since it's not a discriminate union (yet). 

And here we have another union (`DateBucket | NumericBucket`) inside it again, which is also a bit difficult to split. 

I would suggest changing this code, so that we only need to do one splitting of unions, and that is on the `AggregationsResult` level.

```suggestion
export interface DateBucketsAggregationResult {
    buckets: DateBucket[];
}

export interface NumberBucketsAggregationResult {
    buckets: NumericBucket[];
}
```

and then later:

```typescript
export type AggregationsResult = 
  | DateBucketsAggregationResult 
  | NumberBucketsAggregationResult 
  | StatsAggregationResult 
  | SingleValueMetricAggregationResult;
```

**Review by tajakobsen — CHANGES_REQUESTED** [review]
I don't feel I have so much to say regarding correctness of this code. But I have one suggestion related to developer ergonomics that can be considered.

**Comment by edloidas** [comment]
We have a working [test1](https://github.com/enonic/xp/blob/master/modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest.js#L220) and [test2](https://github.com/enonic/xp/blob/master/modules/lib/lib-content/src/test/resources/test/QueryContentHandlerTest_dsl_query.js#L229), where the Stats are used for top level aggregations, so I'm not sure it is for sub-aggregations only:
Maybe @rymsha or @anatol-sialitski can tell more about this.

**Comment by ComLock** [comment]
"The stats-aggregations calculates the following statistics for the parent-aggregation buckets:"
https://developer.enonic.com/docs/xp/stable/storage/aggregations#stats

**Comment by edloidas** [comment]
@ComLock In the files you've mention, StatsAggregation is not present in the results, so maybe those examples must be updated? Yeah, I saw that in docs too, but they must be misleading. Let's wait for more feedback from core devs.

**Comment by ComLock** [comment]
According to ES docs, stats can be a top-level aggregation, so that could be an error in our documentation.

**Comment by rymsha** [comment]
> "The stats-aggregations calculates the following statistics for the parent-aggregation buckets:" https://developer.enonic.com/docs/xp/stable/storage/aggregations#stats

ES can aggregate 
https://www.elastic.co/guide/en/elasticsearch/reference/2.4/search-aggregations-metrics-stats-aggregation.html

So, XP docs are not correct.

**Comment by ComLock** [comment]
Looking at the types, I can see that StatsAggregation, GeoDistanceAggregation, MinAggregation, MaxAggregation and ValueCountAggregation cannot have SubAggregations. We might want to reflect that in the output types aswell?

**Comment by edloidas** [comment]
@ComLock Well, they don't have sub-aggregations if you look at the code. Or am I missing something?

File: modules/lib/lib-content/src/main/resources/lib/xp/content.ts
Comment (edloidas) [reply modules/lib/lib-content/src/main/resources/lib/xp/content.ts:86]: Not sure we can split those buckets. 
@rymsha can we return the mixed buckets, list of both `DateRangeBucket` and `NumericRangeBucket`? It's hard to say from the [mapper](https://github.com/enonic/xp/blob/master/modules/lib/lib-node/src/main/java/com/enonic/xp/lib/node/mapper/AggregationMapper.java#L70)? We can return, at least.

**Comment by ComLock** [comment]
What code, the Java code? No I have not looked at it. But I agree that those probably don't have subAggregations.
Which is reflected in the aggregation input types, but not in the aggregation output types.

Unless one says that the undefined here:
https://github.com/enonic/xp/blob/71cb937e1e88093d8e00324279d0dc03ab874994/modules/lib/lib-content/src/main/resources/lib/xp/content.ts#L66

Is good enough.

TypeScript is not Java so aiming for perfection might not be "cost effective", I think there's such a thing as "good enough" for now and "improve if needed later".

**Comment by edloidas** [comment]
I meant all those types in `node.ts` and `content.ts`. The "input" types you've mentioned, they don't have sub aggregations. But the "output" type `Bucket` indeed have. It'll be risky trying to match input with output, at least for this release, so the "output" have Java-like, more loose type, that is not affected by the input.

**Review by tajakobsen — APPROVED** [review]

**Review by alansemenov — APPROVED** [review]

**Review by rymsha — APPROVED** [review]

**Pull request merged**
